irohad/synchronizer/impl/synchronizer_impl.cpp

```diff
@@ -42,21 +42,39 @@ namespace iroha {
     void SynchronizerImpl::processOutcome(consensus::GateObject object) {
       log_->info("processing consensus outcome");
 
-      visit_in_place(
-          object,
-          [this](const consensus::PairValid &msg) { this->processNext(msg); },
-          [this](const consensus::VoteOther &msg) {
-            this->processDifferent(msg, SynchronizationOutcomeType::kCommit);
-          },
-          [this](const consensus::ProposalReject &msg) {
-            this->processDifferent(msg, SynchronizationOutcomeType::kReject);
-          },
-          [this](const consensus::BlockReject &msg) {
-            this->processDifferent(msg, SynchronizationOutcomeType::kReject);
-          },
-          [this](const consensus::AgreementOnNone &msg) {
-            this->processDifferent(msg, SynchronizationOutcomeType::kNothing);
-          });
+      auto process_reject = [this](auto outcome_type, const auto &msg) {
+        assert(msg.ledger_state->top_block_info.height
+               == msg.round.block_round);
+        notifier_.get_subscriber().on_next(
+            SynchronizationEvent{outcome_type, msg.round, msg.ledger_state});
+      };
+
+      visit_in_place(object,
+                     [this](const consensus::PairValid &msg) {
+                       assert(msg.ledger_state->top_block_info.height + 1
+                              == msg.round.block_round);
+                       this->processNext(msg);
+                     },
+                     [this](const consensus::VoteOther &msg) {
+                       assert(msg.ledger_state->top_block_info.height + 1
+                              == msg.round.block_round);
+                       this->processDifferent(msg, msg.round.block_round);
+                     },
+                     [&](const consensus::ProposalReject &msg) {
+                       process_reject(SynchronizationOutcomeType::kReject, msg);
+                     },
+                     [&](const consensus::BlockReject &msg) {
+                       process_reject(SynchronizationOutcomeType::kReject, msg);
+                     },
+                     [&](const consensus::AgreementOnNone &msg) {
+                       process_reject(SynchronizationOutcomeType::kNothing,
+                                      msg);
+                     },
+                     [this](const consensus::Future &msg) {
+                       assert(msg.ledger_state->top_block_info.height + 1
+                              < msg.round.block_round);
+                       this->processDifferent(msg, msg.round.block_round - 1);
```

Comment:
it would be nice to leave a comment, why `block_round - 1` is used, to avoid a hard time of investigation in the future. something like

> we do not know the ledger state for round `n`, so we cannot claim that the bunch of votes we got is a commit certificate and hence we do not know if the block `n` is committed and cannot require its acquisition.